### PR TITLE
feat!: replace the influx trace client with the local trace client

### LINKED ISSUE #1278: Create a more performant tracer for a network wide trace
Tracing the entire network would likely greatly speed up debugging certain things about gossiping because we would be able to examine the lifecycle of each blockpart, tx, and vote.

When we attempt to trace the entire network with a single influx instance, we cause the network's throughput to decrease substantially. This makes the data collected unreliable. It might be possible to fix the issue by creating multiple influx instances and/or increases the resources of each node, however spinning up multiple influx instances is cumbersome, and we're already pushinght limits of resources that can be alloacted in testground.

Therefore, we should create a performant local only tracer option. This option could work similarly to the current option, but intead of writing to an external database, we would write files locally and then serve them / push them to a server some point after the trace period or the experiment concludes.

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

eventually closes #1278 

super breaking for the config

### CODE REVIEW

File: pkg/trace/local_client.go
Comment (evan-forbes) [inline pkg/trace/local_client.go:138]: by not locking here, we get a lot closer to performance of having individual channels. If we wanted to get a bit more clever, we could additionally split into different channels which serialize and write to files in parallel.

I might do this, along with not blocking when we write to ensure that we never mess with performance. although we might want to keep track of when that occurs.

**Comment by cmwaters** [comment]
Are we planning on deprecating support for influx in favour of this?

File: config/config.go
Comment (cmwaters) [inline config/config.go:1242]: can we call this "trace" by default?

File: pkg/trace/local_client.go
Comment (cmwaters) [inline pkg/trace/local_client.go:54]: Why `cannal`? Is that even a word?

File: pkg/trace/local_client.go
Comment (cmwaters) [inline pkg/trace/local_client.go:46]: nit: generally the mutex should be directly over the fields it protects. If it doesn't protect the `cannal`, then this should be above the mutex

Comment (cmwaters) [reply pkg/trace/local_client.go:46]: Do we even need this mutex if all files are created ahead of time

Comment (cmwaters) [reply pkg/trace/local_client.go:46]: There's also a package called `autofile` which was used by the WAL and has a similar mechanism of creating new files everytime a certain file reaches a size (IIRC)

File: config/config.go
Comment (cmwaters) [inline config/config.go:1243]: I feel like this should be a lot larger than 20. Maybe 100 and then call flush after a proposal is sent (or some other period where we are usually waiting and not blocking the network)

File: pkg/trace/local_client.go
Comment (evan-forbes) [reply pkg/trace/local_client.go:54]: yeah its French for channel

File: mempool/v1/reactor.go
Comment (evan-forbes) [inline mempool/v1/reactor.go:312]: these were moved to only write if the send was successful

File: pkg/trace/local_client.go
Comment (evan-forbes) [inline pkg/trace/local_client.go:169]: got uploading to work I think, but didn't get the server to actually receive it for push or pulling

File: pkg/trace/schema/consensus.go
Comment (evan-forbes) [inline pkg/trace/schema/consensus.go:75]: initially was not planning on removing influx support, but using reflection to convert to map of interfaces seems so silly. We could have kept the huge number of constants, but they are redundant to the struct definitions which is equally bad.

File: pkg/trace/local_client.go
Comment (evan-forbes) [inline pkg/trace/local_client.go:134]: thought about using csv, as that should be faster and create significantly smaller files, however the std lib csv is kinda meh tbh and I didn't want to introduce a new dep. if we're ok with using https://github.com/jszwec/csvutil , then I'm totally down.

File: config/config.go
Comment (evan-forbes) [reply config/config.go]: good catch, put it at 1k, but we might even want to go higher depending on what we're tracing or if we see that we block ever

**Comment by staheri14** [comment]
Could you please include some guidelines on how to utilize the new tracing feature? Alternatively, if this feature is not yet operational within the current PR, an overview of the modifications made would be highly appreciated.

File: pkg/trace/local_client.go
Comment (evan-forbes) [reply pkg/trace/local_client.go]: fixed with new scheme https://github.com/celestiaorg/celestia-core/pull/1279/commits/b77ed97bbb8f8e3cb499696fbec4ce454ba00b19

**Comment by evan-forbes** [comment]
> Alternatively, if this feature is not yet operational within the current PR, an overview of the modifications made would be highly appreciated.

there's currently no way to move the files via a server, however this will trace things by simply changing the default config from 'noop' to 'local', and then selecting tables like normal. 

I deleted the old docs and will include new ones with the server. I can add super short readme here if needed, although I'd like not to block on it so we can get this out asap

**Review by cmwaters — DISMISSED** [review]
LGTM


As I mentioned, if we only want trace data over the span of a height, this entire thing could reside in memory

File: config/toml.go
Comment (cmwaters) [inline config/toml.go:567]: I actually don't think this is breaking in the sense that it will require operators to make any modifications. They will be able to continue running their node as normal.

It is breaking if you had previously been using the influx tracing but I don't think anyone is. If we wanted to be more careful about this we should mention this as an experimental feature

File: mempool/cat/reactor.go
Comment (cmwaters) [inline mempool/cat/reactor.go:263]: This could potentially lead to problems with the visualising tool because it may look like a transaction was sent but never reached it's target but in reality it did reach the target node but was rejected at a later point. Also this doesn't capture the timeliness of the message. For example I want to know the difference between when a transaction was received and when the node sends the `HasTx` message out to its peers

File: mempool/v1/reactor.go
Comment (cmwaters) [reply mempool/v1/reactor.go:312]: Makes sense

File: consensus/reactor.go
Comment (cmwaters) [inline consensus/reactor.go:360]: Do we need to save the height and round if we're saving the entire msg.Vote?

File: consensus/reactor.go
Comment (cmwaters) [inline consensus/reactor.go:341]: Can we also trace when we receive the proposal message?

File: consensus/reactor.go
Comment (evan-forbes) [reply consensus/reactor.go:360]: unrelated to this PR, but I tend to agree. iirc this was to make it easier to find old or weird votes but was not necessary.

wdyt @staheri14? not in this PR, but does it make sense to remove the extra fields and combine the roundstate info we're collecting? we should have a complete view of when votes are sent and received, along with the round state of both nodes

File: consensus/reactor.go
Comment (evan-forbes) [reply consensus/reactor.go:341]: yeah good idea

**Comment by evan-forbes** [comment]
> As I mentioned, if we only want trace data over the span of a height, this entire thing could reside in memory

true good point, however for 100MB block parts we can send and receive a combined 50_000 block parts. We've observed in the past, for example in https://github.com/celestiaorg/celestia-core/issues/1145, that we can see different problematic patterns at different heights. I could also see a scenario where the block proposer happens to be in a weird topology, which causes weird behavior. We might be able to recreate those scenarios in specific tests, however we first have to observe them.

File: consensus/reactor.go
Comment (evan-forbes) [reply consensus/reactor.go:341]: https://github.com/celestiaorg/celestia-core/issues/1285

File: consensus/reactor.go
Comment (staheri14) [reply consensus/reactor.go:360]: >Do we need to save the height and round if we're saving the entire msg.Vote?

The height and round of the vote do not always align with the receiver's height and round. Here, the `height` and `round` represent the receiving node's info, but not the vote.  

>wdyt @staheri14? not in this PR, but does it make sense to remove the extra fields and combine the roundstate info we're collecting? we should have a complete view of when votes are sent and received, along with the round state of both nodes

I support maintaining these extra field as it is simpler and more precise. If we don't retain this information, we would have to depend on the timestamps of the recorded data to establish correlations (except timestamps, is there any other way to correlate the two i.e., round state and recorded votes?).

Comment (cmwaters) [reply consensus/reactor.go:360]: > The height and round of the vote do not always align with the receiver's height and round

That's a good point. Okay you've convinced me. Let's keep it.

File: config/config.go
Comment (staheri14) [reply config/config.go:1237]: As a reminder https://github.com/celestiaorg/celestia-core/issues/1288

File: pkg/trace/local_client.go
Comment (staheri14) [inline pkg/trace/local_client.go:37]: What does `retuen` mean?

File: pkg/trace/local_client.go
Comment (staheri14) [inline pkg/trace/local_client.go:47]:
```suggestion
	// fileMap maps tables to their open files, files are thread-safe, but the map
```

**Review by staheri14 — COMMENTED** [review]
I've managed to review about half of the PR so far, and here are my comments based on what I've seen up to this point. Looks good in general.

File: pkg/trace/local_tracer.go
Comment (staheri14) [inline pkg/trace/local_tracer.go:144]: Using JSON format can be inefficient because it repeats field names as keys for every event, leading to redundancy. Switching to CSV could be more efficient.

Comment (evan-forbes) [reply pkg/trace/local_tracer.go:144]: yeah csv is definitely better, and should be a simple enough switch see

https://github.com/celestiaorg/celestia-core/pull/1279#discussion_r1550497217

File: pkg/trace/local_client.go
Comment (evan-forbes) [reply pkg/trace/local_client.go]: mispelled return

File: pkg/trace/local_client.go
Comment (evan-forbes) [reply pkg/trace/local_client.go]: mispelled return alot :sweat_smile:

File: pkg/trace/tracer.go
Comment (evan-forbes) [reply pkg/trace/tracer.go:26]: I believe nodeID is refered to interchangably with peerID, so the public key of the key used to sign gossiped messages https://github.com/celestiaorg/celestia-core/blob/abc516304122da41e154b49d0a152360e38736a8/p2p/peer.go#L26

Comment (staheri14) [reply pkg/trace/tracer.go:26]: I think it's worth adding a comment about it to the func description.

Comment (evan-forbes) [reply pkg/trace/tracer.go:29]: sounds good! https://github.com/celestiaorg/celestia-core/pull/1279/commits/4439d32cf2c737922fe5cadf2a009ec9429a9bfd

File: config/config.go
Comment (evan-forbes) [reply config/config.go]: good catch! removed the no longer needed validation

Comment (evan-forbes) [reply config/config.go]: https://github.com/celestiaorg/celestia-core/pull/1279/commits/779a68bb35f146db906b887fde771235f1e7a529

File: config/config.go
Comment (cmwaters) [inline config/config.go:1275]: Can it be 0? Do we not want `<=`

**Review by cmwaters — APPROVED** [review]
lgtm (looks like there is one failing test)

File: config/config.go
Comment (evan-forbes) [reply config/config.go:1275]: technically speaking, 0 could work although ofc that's a bad idea. The reasoning for switching here was so that a value of 0 won't fail for when the default value isn't used in some tests, and that I wasn't sure where to stop. While 0 is not reccomended, neither is 1. combining those two things resulted in just switching this to 0

File: pkg/trace/schema/mempool.go
Comment (staheri14) [inline pkg/trace/schema/mempool.go:24]: [nit] 
```suggestion
// MempoolTx describes the schema for the "mempool_tx" table.
```

File: pkg/trace/README.md
Comment (staheri14) [inline pkg/trace/README.md:19]: [nit] Based on my testing, this field automatically gets populated with the names of all available tables hence, I suppose there is no need for modifications unless there's a specific desire to exclude certain tables from the tracing process. If that's the case, it would indeed be helpful to clarify that it is possible to selectively skip tracing specific tables.

**Review by staheri14 — APPROVED** [review]
LGTM! Thanks for updating the README. I conducted a test, which allowed me to gather locally traced data. Also, I reviewed some of this data using the `DecodeFile` function. Left some non-blocking comments.

File: pkg/trace/schema/mempool.go
Comment (evan-forbes) [reply pkg/trace/schema/mempool.go:24]: I'll grab this in a followup, thanks

File: pkg/trace/README.md
Comment (evan-forbes) [reply pkg/trace/README.md:19]: thanks for not blocking on this

I think I'm having some author blindness here :sweat_smile:, so I'll defer to you on this. are you refering to the readme, or do you think the first sentence here doesn't describe that? if its the first sentence, do you think 

> The list of tables that are updated when tracing. Tables not included below will not be traced.

works better?

**Comment by evan-forbes** [comment]
also, thanks @staheri14 for manually trying out the code!!! :smiley:

**Pull request merged**
